zerocloud/proxyquery.py

```diff
@@ -1694,10 +1712,15 @@ def _process_response(self, conn, request):
         return self.process_server_response(conn, request, resp)
 
     def _connect_exec_node(self, obj_nodes, part, request,
-                           logger_thread_locals, cnode, request_headers):
+                           logger_thread_locals, cnode, request_headers,
+                           known_nodes, salt):
         self.app.logger.thread_locals = logger_thread_locals
         conn = None
-        for node in obj_nodes:
+        for node in chain(known_nodes, obj_nodes):
+            if (known_nodes and node in known_nodes) \
+                    or (not known_nodes and cnode.location):
```

Comment:
if we have `known_nodes` it means that we are co-locating, and previous the job was sent to `known_nodes[0]`. Therefore first expression means: we are co-locating and currently trying to send to node that was already used (which have a good chance to succeed).
Second expression means: we are co=locating, because `cnode.location` is defined but known_nodes is not set, therefore this node is the first one we trying to co-locate on (if success this node will be in the `known_nodes` next time.